src/kv/RocksDBStore.cc

```diff
@@ -485,6 +485,12 @@ int RocksDBStore::load_rocksdb_options(bool create_if_missing, rocksdb::Options&
   if (cct->_conf.get_val<Option::size_t>("rocksdb_metadata_block_size") > 0)
     bbt_opts.metadata_block_size = cct->_conf.get_val<Option::size_t>("rocksdb_metadata_block_size");
 
+  // Set max_total_wal_size if unset
+  // This is only used if we have some non-default column families,
+  // But value rocksdb uses by default is way too large
+  if (opt.max_total_wal_size == 0) {
+    opt.max_total_wal_size = opt.write_buffer_size;
```

Comment:
Isn't it better to adjust bluestore_rocksdb_options parameter and insert proper vallue for max_total_wal_size there?
BTW I can see that write_buffer_size is set to 256MB in Ceph by default...

Reply:
I modified a bit store_test (https://github.com/ceph/ceph/compare/master...aclamk:dnm-test-wal-sharding) to test different settings of max_total_wal_size.

Test procedure:
```bin/ceph_test_objectstore --gtest_filter=*SpilloverTest*/2 --bluestore_rocksdb_options "max_total_wal_size=${wal},compression=kNoCompression,max_write_buffer_number=4,min_write_buffer_number_to_merge=1,recycle_log_file_num=4,writable_file_max_buffer_size=0,compaction_readahead_size=2097152" --bluestore_rocksdb_cf true --no-log-to-stderr 2>/dev/null |grep WAL; ```

Results:
```
WAL=200000000
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES       
WAL         0 B         384 MiB     0 B         0 B         0 B         383 MiB     2           
WAL         0 B         384 MiB     0 B         0 B         0 B         383 MiB     
WAL=400000000
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES       
WAL         0 B         1.4 GiB     138 MiB     0 B         0 B         445 MiB     8           
WAL         0 B         1.6 GiB     138 MiB     0 B         0 B         684 MiB     
WAL=600000000
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES       
WAL         0 B         1.5 GiB     51 MiB      0 B         0 B         659 MiB     11          
WAL         0 B         1.6 GiB     51 MiB      0 B         0 B         818 MiB     
WAL=800000000
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES       
WAL         0 B         1.3 GiB     28 MiB      0 B         0 B         547 MiB     10          
WAL         0 B         1.6 GiB     56 MiB      0 B         0 B         851 MiB     
WAL=1000000000
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES       
WAL         0 B         1.3 GiB     33 MiB      0 B         0 B         571 MiB     11          
WAL         0 B         1.5 GiB     197 MiB     0 B         0 B         970 MiB     
WAL=1500000000
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES       
WAL         0 B         1.1 GiB     33 MiB      0 B         0 B         567 MiB     10          
WAL         0 B         1.4 GiB     173 MiB     0 B         0 B         1.4 GiB     
WAL=2000000000
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES       
WAL         0 B         1.1 GiB     38 MiB      0 B         0 B         567 MiB     10          
WAL         0 B         1.9 GiB     150 MiB     0 B         0 B         1.9 GiB     
WAL=2500000000
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES       
WAL         0 B         1.2 GiB     61 MiB      0 B         0 B         547 MiB     10          
WAL         0 B         2.2 GiB     146 MiB     0 B         0 B         2.3 GiB     
WAL=3000000000
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES       
WAL         0 B         1.2 GiB     44 MiB      0 B         0 B         567 MiB     10          
WAL         0 B         2.2 GiB     637 MiB     0 B         0 B         2.8 GiB
```

It seems that if WAL is smaller then 1GB rocksdb has trouble keeping it below set value.
For WAL 1GB and larger rocksdb does not go over limit.

Reply:
@aclamk - you might want to increase object count for the test case. Here are my results for 4K objects and 1GB WAL:
CFamily on, 4096 obj, 1GB wal
1 : device size 0xc0000000 : own 0x[2000~bfffe000] = 0xbfffe000 : using 0xbc7fe000(2.9 GiB)
2 : device size 0x105aa15000 : own 0x[10000~3e03d0000,7d9940000~a7780000] = 0x487b50000 : using 0x454c20000(17 GiB) : bluestore has 0xbd2eb0000(47 GiB) available
RocksDBBlueFSVolumeSelector: wal_total:0, db_total:3060164198, slow_total:66727998054, db_avail:0
Usage matrix:
DEV/LEV     WAL         DB          SLOW        *           *           REAL        FILES
LOG         0 B         16 MiB      0 B         0 B         0 B         11 MiB      1
WAL         0 B         2.8 GiB     271 MiB     0 B         0 B         1.2 GiB     7
DB          0 B         177 MiB     2.1 GiB     0 B         0 B         2.3 GiB     30
SLOW        0 B         0 B         15 GiB      0 B         0 B         15 GiB      229
TOTALS      0 B         2.9 GiB     17 GiB      0 B         0 B         0 B         267
MAXIMUMS:
LOG         0 B         16 MiB      0 B         0 B         0 B         11 MiB
WAL         0 B         2.8 GiB     271 MiB     0 B         0 B         1.6 GiB
DB          0 B         1.7 GiB     3.1 GiB     0 B         0 B         3.6 GiB
SLOW        0 B         0 B         15 GiB      0 B         0 B         15 GiB
TOTALS      0 B         3.0 GiB     18 GiB      0 B         0 B         0 B

Please note WAL maximum at 1.6GiB

Reply:
And in results I shared for 256MB WAL it never exceeds 512+ MB which is inline with what I ever have observed  before. Personally I'd prefer to preserve these actual numbers. Not many  pros and cons though. Just prefer being conservative here as one should be in data storage solutions ;)

Reply:
@ifed01 So, basically, you proposed to give default of 256MB, but document that it is likely to grow as much as 512?

Reply:
Originally I didn't imply any documentation on that and moreover I'm not 100% sure it is always below 512MB. So may be it's better to simply avoid this sort of information in log. Just recommend 4GB for WAL volume and omit rationales for that.